Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -515,6 +479,159 @@ public final class SourceKitServer: LanguageServer {
   }
 }
 
+// MARK: - MessageHandler
+
+extension SourceKitServer: MessageHandler {
+  public func handle<N: NotificationType>(_ params: N, from clientID: ObjectIdentifier) {
+    queue.async {
+
+      let notification = Notification(params, clientID: clientID)
+      self._logNotification(notification)
+
+      switch notification {
```

Comment:
I don't know if you've benchmarked this, but it might be worth comparing this to switching over the method string.  Dynamic casting can be expensive.

Reply:
I just benchmarked it with the following test and each check takes about 350ns in debug builds. When matching the first case, it takes 224ns. I think that’s quite negligible.

<details><summary>Here’s the test</summary>
<p>

```swift
import Foundation

public final class Request<R> {}

struct S1 {}
struct S2 {}
struct S3 {}
struct S4 {}
struct S5 {}
struct S6 {}
struct S7 {}
struct S8 {}
struct S9 {}
struct S10 {}
struct S11 {}
struct S12 {}
struct S13 {}
struct S14 {}
struct S15 {}
struct S16 {}
struct S17 {}
struct S18 {}
struct S19 {}
struct S20 {}
struct S21 {}
struct S22 {}
struct S23 {}
struct S24 {}
struct S25 {}
struct S26 {}
struct S27 {}
struct S28 {}
struct S29 {}
struct S30 {}

protocol X {
  func test<T>(_ x: Request<T>)
}

var output: String = ""

struct XX: X {
  func test<T>(_ x: Request<T>) {
    switch x {
    case let x as Request<S1>:
      output = "S1"
    case let x as Request<S2>:
      output = "S2"
    case let x as Request<S3>:
      output = "S3"
    case let x as Request<S4>:
      output = "S4"
    case let x as Request<S5>:
      output = "S5"
    case let x as Request<S6>:
      output = "S6"
    case let x as Request<S7>:
      output = "S7"
    case let x as Request<S8>:
      output = "S8"
    case let x as Request<S9>:
      output = "S9"
    case let x as Request<S10>:
      output = "S10"
    case let x as Request<S11>:
      output = "S11"
    case let x as Request<S12>:
      output = "S12"
    case let x as Request<S13>:
      output = "S13"
    case let x as Request<S14>:
      output = "S14"
    case let x as Request<S15>:
      output = "S15"
    case let x as Request<S16>:
      output = "S16"
    case let x as Request<S17>:
      output = "S17"
    case let x as Request<S18>:
      output = "S18"
    case let x as Request<S19>:
      output = "S19"
    case let x as Request<S20>:
      output = "S20"
    case let x as Request<S21>:
      output = "S21"
    case let x as Request<S22>:
      output = "S22"
    case let x as Request<S23>:
      output = "S23"
    case let x as Request<S24>:
      output = "S24"
    case let x as Request<S25>:
      output = "S25"
    case let x as Request<S26>:
      output = "S26"
    case let x as Request<S27>:
      output = "S27"
    case let x as Request<S28>:
      output = "S28"
    case let x as Request<S29>:
      output = "S29"
    case let x as Request<S30>:
      output = "S30"
    default:
      output = "other"
    }
  }
}


func main(_ x: X) {
  let a = Request<S1>()
  let start = Date()
  let iterations = 1_000_000_0
  for _ in 0..<iterations {
    x.test(a)
  }
  print("\(Date().timeIntervalSince(start) * 1000 * 1000 * 1000 / Double(iterations))ns")
}


main(XX())
```

</p>
</details>